### PR TITLE
Avoid BuildHost crash in Mono due to missing types

### PR DESCRIPTION
**Pull request by uweigand**
The LoadProjectFileAsync routine calls EnsureMSBuildLoaded to make sure the Microsoft.Build.dll is accessible, but requires types from that assembly within the routine itself.  This may not always work with Mono, as the JIT may look up the types during compilation.

Fixed by splitting out a LoadProjectFileCoreAsync routine, similarly to what is already done for GetProjectsInSolution.

Fixes https://github.com/dotnet/runtime/issues/101121.

### CODE REVIEW

**Review by dibarbet — APPROVED** [review]
I'm fine with this, but would like @jasonmalinowski to take a look as well.

**Review by JoeRobich — APPROVED** [review]

File: src/Workspaces/Core/MSBuild.BuildHost/BuildHost.cs [hunk 1]
```diff
@@ -161,9 +161,15 @@ private static ImmutableArray<(string ProjectPath, string ProjectGuid)> GetProje
...
     /// Returns the target ID of the <see cref="ProjectFile"/> object created for this.
     /// </summary>
-    public async Task<int> LoadProjectFileAsync(string projectFilePath, string languageName, CancellationToken cancellationToken)
+    public Task<int> LoadProjectFileAsync(string projectFilePath, string languageName, CancellationToken cancellationToken)
     {
         EnsureMSBuildLoaded(projectFilePath);
+        return LoadProjectFileCoreAsync(projectFilePath, languageName, cancellationToken);
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)] // Do not inline this, since this uses MSBuild types which are being loaded by the caller
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/MSBuild.BuildHost/BuildHost.cs:170]: You may want to make a note that this was mono specific, lest somebody tries to undo this change being confused why it works on other platforms.

**Review by jasonmalinowski — APPROVED** [review]

**Comment by jasonmalinowski** [comment]
I filed https://github.com/dotnet/roslyn/issues/74407 to look at what it'd take to get our tests running on mono to catch issues like this.

**Comment by uweigand** [comment]
Not sure why this is failing now after just a comment change ... seems an infrastructure issue.

**Comment by jaredpar** [comment]
The test run failed. Think it's still using stale integration test bits. Worry that this PR may be snagged on stale state. I rebased locally, pushed to a new branch and opened a new PR. Hopefully that one will get past these issues #74994.

**Comment by jaredpar** [comment]
Closing as I got this change to merge in cleanly in #74994. 

Sorry for losing track of this.
